src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs

```diff
@@ -95,28 +85,121 @@ public class PublishPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
+                    throw new InvalidOperationException(
+                        StringUtil.Loc("CannotUploadFromCurrentEnvironment", hostType ?? string.Empty)); 
+                }
+
+                if (String.IsNullOrWhiteSpace(artifactName))
+                {
+                    string jobIdentifier = context.Variables.GetValueOrDefault(WellKnownDistributedTaskVariables.JobIdentifier).Value;
+                    var normalizedJobIdentifier = NormalizeJobIdentifier(jobIdentifier);
+                    artifactName = normalizedJobIdentifier;
+                }
+
+                if(!PipelineArtifactPathHelper.IsValidArtifactName(artifactName)) {
+                    throw new ArgumentException(StringUtil.Loc("ArtifactNameIsNotValid", artifactName));
+                }
+
+                string fullPath = Path.GetFullPath(targetPath);
+                bool isFile = File.Exists(fullPath);
+                bool isDir = Directory.Exists(fullPath);
+                if (!isFile && !isDir)
+                {
+                    // if local path is neither file nor folder
+                    throw new FileNotFoundException(StringUtil.Loc("PathDoesNotExist", targetPath));
+                }
+
+                // Upload to VSTS BlobStore, and associate the artifact with the build.
+                context.Output(StringUtil.Loc("UploadingPipelineArtifact", fullPath, buildId));
+                PipelineArtifactServer server = new PipelineArtifactServer();
+                await server.UploadAsync(context, projectId, buildId, artifactName, fullPath, token);
+                context.Output(StringUtil.Loc("UploadArtifactFinished"));
 
-            string fullPath = Path.GetFullPath(targetPath);
-            bool isFile = File.Exists(fullPath);
-            bool isDir = Directory.Exists(fullPath);
-            if (!isFile && !isDir)
-            {
-                // if local path is neither file nor folder
-                throw new FileNotFoundException(StringUtil.Loc("PathDoesNotExist", targetPath));
             }
+            else if (artifactType == fileShareType)
+            {
+                string fileSharePath = context.GetInput(ArtifactEventProperties.FileSharePath, required: true);
+                string artifactPath = Path.Join(fileSharePath, artifactName);
+
+                if(RuntimeInformation.IsOSPlatform(OSPlatform.Windows)){
+                    // create the artifact. at this point, mkdirP already succeeded so the path is good.
+                    // the artifact should get cleaned up during retention even if the copy fails in the
+                    // middle
+                    Directory.CreateDirectory(artifactPath);
+
+                    // 2) associate the pipeline artifact with an build artifact
+                    VssConnection connection = context.VssConnection;
+                    BuildServer buildHelper = new BuildServer(connection);
```

Comment:
>buildHelper [](start = 32, length = 11)

buildServer

Reply:
This is really a naming convention in the agent but I'm not against naming it buildServer either.